server/src/instant/grouped_queue.clj

```diff
@@ -0,0 +1,53 @@
+(ns instant.grouped-queue
+  (:import
+   (java.util.concurrent LinkedBlockingQueue)
+   (clojure.lang PersistentQueue)))
+
+(def empty-q PersistentQueue/EMPTY)
+
+(defn create [group-fn]
+  {:group-fn group-fn
+   :group-key->subqueue (atom {})
+   :main-queue (LinkedBlockingQueue.)})
+
+(defn enqueue! [{:keys [group-fn main-queue group-key->subqueue]
+                 :as _grouped-q} item]
+  (let [group-key (group-fn item)]
+    (if (nil? group-key)
+      ;; This item is not to be grouped.
+      (.put main-queue [:item item])
+      ;; This item will be grouped on `group-key` 
+      (let [[prev] (swap-vals! group-key->subqueue
```

Comment:
I am using 1 atom, rather than an atom per queue, for the following reason: 

If we wanted to keep `group-key->subqueue` as introspectable state, we would have to keep it in something like an atom. But then, if each subqueue also was an atom, we would need to coordinate changes between the two atoms.

My thinking is since the contention here is strictly for adding and removing elements, it's okay to have everything in one atom. 

Open to other ideas too though!